PM-3133: Block executor

aakoshh opened the pull request:
Added a `BlockExecutor` component to farm out the `ExecuteBlocks` effects from the `ConsensusService`. It has an internal work queue where batches are added. A successful execution is noted down in the `ViewStateStorage`. The actual execution is delegated to the `ApplicationService`. If there's an error, we try again next time there's a new batch.

If the `ConsensusService` and the `SyncService` skip ahead to a newer Commit Quroum Certificate, and the last executed block is no longer available, the `BlockExecutor` skips blocks too. Not much else it can do. These events are traced.

rtkaczyk commented on metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/storage/BlockStorageProps.scala:
```diff
@@ -238,6 +254,53 @@ object BlockStorageProps extends Properties("BlockStorage") {
+      ancestor = prefix.last
+      postfix    <- genNonEmptyBlockTree(ancestor.id)
+      descendant <- Gen.oneOf(postfix)
+      data = TestData(prefix ++ postfix)
```
Do we need the ancestors of the `ancestor`? Why not just generate a single block tree?

aakoshh replied:
I wanted to make sure that I can pick any 2 blocks in the tree which are related, and that the traversal will only go between them, not higher to the root for example. 

Conceptually I only need a single block tree, but the generator only returns a list of blocks, it's not obvious how to pick an ancestor and a descendant from that, unless one of them is the root. By doing it in 2 stages, I know that every block in the `postfix` is a descendant of the `ancestor`, so I can pick any of them, and `ancestor` may or may not be the `root` of the `prefix`. 

I figured it's easier to generate the data in the most convenient way, so there doesn't need to be any logic in the test.

rtkaczyk replied:
> that the traversal will only go between them, not higher to the root for example.

Makes sense, but we're not really testing that, are we? I mean we're only looking at the result, not at how it's produced. In fact, maybe it could actually test the traversal if the ancestor was always the root, as it would produce an error if we try to access its parent?

Not a big deal, just pointing out

aakoshh replied:
For a few minutes I didn't know what you meant by not testing it. Now I understand you meant that it _could_ have accessed the parent during traversal, just not returned it. 

Sorry, I didn't mean to check what the traversal does behind the scenes, only that we return the `ancestor` as the `head` of `path` (which we [check](https://github.com/input-output-hk/metronome/pull/39/files/7fd73c729ad087906359aa78ad637b9ce3b41108#diff-040272a08fe005f889fce95f80d70b6c857e6cf9a776c14a0fd001c98628a9fcR274-R275)), not just by accident because it doesn't have a parent to go to, but because that's what the result is supposed to be. 

The `ancestor` _can_ be the root of the prefix (if it's a single element tree), so if it threw an exception trying to access its parent it should fail the test in some cases. I assume you didn't want to suggest by "always the root" that this should be the only way we should test it?

rtkaczyk replied:
Yeah, makes sense 👍

rtkaczyk commented on metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutorProps.scala:
```diff
@@ -0,0 +1,287 @@
+    }
+  }
+
+  property("executeBlocks - from last") = forAll { (fixture: TestFixture) =>
```
Should we set `maxBatches = 1` since we're only enqueuing the last one?

rtkaczyk replied:
Sorry, but I fail to see how the other batches are relevant here:

> at the time it takes on the next batch

```
_ <- blockSychronizer.enqueue(lastBatch)
```

Since we're executing a single batch, there is no next batch. Am I missing something?

aakoshh replied:
No, you're not missing anything, it's only using the last batch, the others may only be there to emphasise that their presence should make no difference. 

Sorry if I didn't make myself clear: I agree with you that setting `maxBatches = 1` would achieve the same effect. I just didn't want to specify something that doesn't have any bearing on whether the test should work, to avoid making it look like that's a crucial setting (like it is when I set `minBatches = 2` in the failure case). The only effect should be that test generation may become quicker.

Now, if there wasn't the `prefixTree` in the data generation (a 0th batch if you will), then it could have an effect, because in that case `maxBatches = 1` would mean always going from the root.

rtkaczyk replied:
> test generation may become quicker.

That's what I was after

> avoid making it look like that's a crucial setting

but I'm fine with this reasoning

rtkaczyk commented on metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutorProps.scala:
```diff
@@ -0,0 +1,287 @@
+        val lastExecutedBlockHash = lastBatch.lastExecutedBlockHash
+        for {
+          _ <- fixture.storeRunner.runReadWrite {
+            TestBlockStorage.pruneNonDescendants(lastExecutedBlockHash)
```
I don't quite understand this test. Do we even populate the storage with such non-descendants?

aakoshh replied:
Sorry, there aren't many comments in the setup.

Yeah, we generate a [prefixTree](https://github.com/input-output-hk/metronome/pull/39/files/7fd73c729ad087906359aa78ad637b9ce3b41108#diff-43a00c2f1553bcab8cd7e482b1c3d5861d3cc10c91528c904910a85a46e909f4R176) first and then between 1 and 5 extension trees; this test picks `lastBatch.lastExecutedBlockHash` as the new root, so after pruning we shall only have the last extension tree. The non-descendants are the ancestors, uncles and cousins in the prefix tree and any previous extensions.

The test checks that even if we prune the block that were the last executed ones according to our view state, the `BlockExecutor` can recover by going from whatever the last committed block before the current batch was.

This is implemented [here](https://github.com/input-output-hk/metronome/pull/39/files/7fd73c729ad087906359aa78ad637b9ce3b41108#diff-6c7370a74c3203456a648dd0b2eb746f549b452dd00006152771767883fba0dbR111-R114) but I can't remember why I thought a situation like this would arise. I only implemented pruning for the robot and it prunes blocks [relative to the last executed hash](https://github.com/input-output-hk/metronome/pull/42/files#diff-10b56ac2536b604ca590b5fe6859da38983d0023e5199eb3d5659d1d9cee12ddR419-R421), so this shouldn't happen. Maybe it was before I added the reading of the last hash from the view state before each batch :thinking:

aakoshh replied:
Yeah, that's how it [started](https://github.com/input-output-hk/metronome/pull/39/commits/50493682ce2decf005352e772bdb178c8529aabb#diff-6c7370a74c3203456a648dd0b2eb746f549b452dd00006152771767883fba0dbR43-R56), and I think I was just being defensive, thinking what would happen if someone would prune away the blocks we thought we last executed, surely they would keep at least the last committed one.

aakoshh replied:
![plans within plans](https://memegenerator.net/img/instances/74103097.jpg)

rtkaczyk replied:
I'm still not quite clear. If I understand correctly the initial store is a chain of blocks, so what is there to be pruned?

aakoshh replied:
The initial store has all the block tree that was generated, and the batches are what will tell what the committed chain from that is. 

Say it has a tree like this were `0*` is the `prefixTree`, `1*` is the first and `2*` is the second extension:
```
                            15
                          /
01 - 02 - 04 - 11 - 12 - 14 - 16 
   \                 \ 
    03                13 - 17 - 21 - 22 - 23
                                      \
                                       24 - 25
```
Then the fixture would have 2 batches: `(04, 17)` and `(17, 25)`. In each pair the 1st member is what has been committed in the previous batch, the 2nd is the new committed block. The pruning in this case would get rid of anything before `17`. 

> the call to pruneNonDescendants is kind of arbitrary. What if it's called with a different blockHash than what we expect?

You are right, the pruning here is specific to one of the conditions that the `BlockExecutor` knows how to handle. If we pruned from `13`, it wouldn't execute anything. The batches mean that we only have commits for `04`, `17` and `25`. The recovery mechanism in the executor works by assuming that even if something went wrong during the execution, the federation will keep working and give us a new commit at some point, at which point we can again try and see if we can find that block. 

In our example, the `lastExecutedBlockHash` would be `01` but for some reason someone pruned to `13`, so the batch of `(04, 17)` would fail because neither `01` nor `04` exist. But then would come batch `(17, 25)` and we would see that `17` indeed exists in storage. We ask the `ApplicationService` to execute `21`, `22`, `24` and `25`, because it always assumes that the first item of the batch has been done before. If `21` fails, then we are stuck, but if we indeed synced to `17` somehow, then it can resume from there  and update `lastExecutedBlockHash`.

The failure scenario is tested further down. This test only checks that if the `lastExecutedHash` in `ViewStateStorage` doesn't exist but the one in the batch does, we use that one to attempt to execute from. 

> maybe we should add a test for the executionSkipped case?

Yeah, good idea, will add one :+1:

aakoshh replied:
Added the test for `executionSkipped` in https://github.com/input-output-hk/metronome/pull/39/commits/40d1a00a78120627012c88cee222526062bb5a8b

Thank you for raising it, it caught a bug as well, the first block that failed the compare-and-set wasn't traced.

rtkaczyk commented:
I feel we've discussed the main code sufficiently when discussing PM-3108/3110, and I'm happy with it 👍 , so I'm focusing my review on the tests

aakoshh commented:
@rtkaczyk on the point of skipped executions I got into a dilemma about this scenario:
1. The block executor starts executing a batch `A`, `B`, `C` and `D`.
2. Things are going slow, currently executing `B`.
3. Meanwhile, the consensus gets out of sync and decides it needs to jump ahead.
4. The federation is on block `F`, we sync there by downloading the state of `F`.
5. Race condition happens: `B` is finished and set as the last executed block hash, then `F` is set as the last executed block hash, but only after the executor set `B`, so the executor carries on with the next block.
6. The executor executes `C`, tries to set the last hash to `C` but realises it's not `B` any more, so it skips `D`. 

Executing `C` without realising that we are already in state `F` could be bad. My first thought was that this can be handled by the `ApplicationService`: 
* It can use a `Semaphore` to make sure it's not doing `stateSync` and `executeBlock` at the same time. However that doesn't solve the above issue alone.
* It can maintain a `lastExecutedBlockHashRef` of its own and check in `executeBlock` that `block.parentHash == lastExecutedBlockHash`, and also set it in `stateSync`. 

I thought this is too much lore to get right, so I moved the call to `stateSync` and the subsequent `fastForwardStorage` from `SyncService` to the `BlockExecutor`, and added a semaphore around the critical sections. This should make sure that we would not be executing `C` because first we execute `A`,`B`,`C` and `D` completely before syncing to `E`.

I know you said the `BlockExecutor` does too much, and now it does even more, but I reckon the overall semantics are easier to follow this way.

I also modified `ApplicationService.syncState` to return a flag to indicate whether it's okay to prune away old blocks, although this could be a constructor argument as well.

rtkaczyk commented on metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutor.scala:
```diff
@@ -0,0 +1,304 @@
+  ): F[Unit] = {
+    val blockHash = Block[A].blockHash(block)
+
+    val prune = for {
```
1. Why are we not using `blockStorage.pruneNonDescendants` here? Is it because there may be no link between the root and the `blockhash`? 
2. What about the safe deletion order?
3. Should we maybe put this into `BlockStorage` as an alternative pruning method so that the difference in use cases can be more easily spotted?

aakoshh replied:
That got me scratching my head, I should add a comment. 

1. You're exactly right, when `syncState` is invoked we can expect that there will be a gap between what we have in the block storage and the block to which we just synced, so there is (probably) no way to traverse to the root from the new block. That's why I added the root block hash explicitly to the view state, so we can go from the other direction. The `skip` is used purely as a safeguard.
2. Arguably at this point we don't care about the deletion order because we assume we are dealing with a forest, and we just want to purge the old tree and keep the new (single) tree. It will be executed as one batch against RocksDB. 
3. The thing is that the block store doesn't store the root, we rely on the view store here to figure out where to go from. We discussed earlier that the block store _can_ store a forest, and it's up to the user to keep track of pointers into each tree. The view store knows a bit more, we know we only care about this one tree. We could move a method that just purges all descendants of a certain block, but there I would try to be more careful about the deletion order as I could not assume the block is the root... maybe a method that goes up to the root and kills the root, a `purgeTree`. Should we do that?

aakoshh replied:
Added in https://github.com/input-output-hk/metronome/pull/39/commits/0b8b29d90c4c45ee2556482664880d0bc0d210cf
Thanks for the suggestion!

rtkaczyk replied:
I like it better that way :+1:

rtkaczyk commented:
It was on my list to look for race conditions regarding `lastExecutedBlockhash`, but it seems you fixed it before I got a change. Cool 🎉

rtkaczyk reviewed (approved):
Just so you know, I've made my peace with the BlockExecutor being what it is. It's certainly not easy to grasp, but should make implementing applications easier 💯
